draft-ietf-quic-recovery.md

```diff
@@ -767,20 +767,20 @@ While in slow start, QUIC increases the congestion window by the
 number of bytes acknowledged when each acknowledgment is processed, resulting
 in exponential growth of the congestion window.
 
-QUIC exits slow start upon loss or upon increase in the ECN-CE counter.
-When slow start is exited, the congestion window halves and the slow start
-threshold is set to the new congestion window.  QUIC re-enters slow start
-any time the congestion window is less than the slow start threshold,
-which only occurs after persistent congestion is declared.
+QUIC MUST exit slow start and enter congestion avoidance upon loss or upon
+increase in the ECN-CE counter. When slow start is exited, the congestion
+window halves and the slow start threshold is set to the new congestion
+window.  QUIC re-enters slow start any time the congestion window is less
+than the slow start threshold, which only occurs after persistent congestion
+is declared.
 
 ## Congestion Avoidance
 
-Slow start exits to congestion avoidance.  Congestion avoidance uses an
-Additive Increase Multiplicative Decrease (AIMD) approach that increases
-the congestion window by one maximum packet size per congestion window
-acknowledged.  When a loss or ECN-CE marking is detected, NewReno halves
-the congestion window, sets the slow start threshold to the new
-congestion window, and then enters the recovery period.
+Congestion avoidance uses an Additive Increase Multiplicative Decrease (AIMD)
+approach that MUST increase the congestion window by one maximum packet size
+per congestion window acknowledged.  When a loss or ECN-CE marking is
```

Comment:
```suggestion
Congestion avoidance uses an Additive Increase Multiplicative Decrease (AIMD)
approach and MUST increase the congestion window by one maximum packet size
per congestion window acknowledged.  When a loss or ECN-CE marking is
```

Reply:
This is a weird MUST to add.  It just does this.

Reply:
Agreed, we can just drop this part of the change.

Reply:
What are we dropping? My suggestion or the MUST increase entirely? Dropping my suggestion is fine, but I would keep the MUST increase -- that's just the definition of Reno.

Reply:
The removal of the first sentence is fine.  (Though we need a definition of congestion avoidance, this isn't it.)

The original text here is fine without a MUST: "Congestion avoidance uses an Additive Increase Multiplicative Decrease (AIMD) approach that increases the congestion window by one maximum packet size per congestion window".  You don't need normative language, because it is definitional.  If you use "MUST" you do so by identifying an actor and an action.  This is saying what "congestion avoidance" is.

Reply:
What I meant to say is that this is the definition of it. Congestion avoidance is simply the state of the controller where `cwnd > ssthresh`, this does not imply that the controller needs to increase by 1 packet size per round trip time. Reno isn't defined anywhere formally as a term, but it is an AIMD controller that "increases the congestion window by 1 packet size per round-trip time" and "decreases it by 0.5 when loss is encountered." My point was that to make our controller Reno, these two behaviors have to be MUSTs. Everything else is window dressing.

Reply:
Now that I look at RFC 5681 a bit more carefully, we probably want to use language similar to what's used there:
```
   During congestion avoidance, cwnd is incremented by roughly 1 full-
   sized segment per round-trip time (RTT).  Congestion avoidance
   continues until congestion is detected.  The basic guidelines for
   incrementing cwnd during congestion avoidance are:

      * MAY increment cwnd by SMSS bytes

      * SHOULD increment cwnd per equation (2) once per RTT

      * MUST NOT increment cwnd by more than SMSS bytes
```

Reply:
The important bit there is the MUST NOT. I don't think the MAY is useful, but the second and third statements are probably adequate.

Reply:
Yes, a MUST NOT would be useful.

Reply:
I would prefer to refer to RFC5681 rather than respecifying these normative requirement here.

Reply:
@mirjak -- this document does not make normative references to RFC 5681.

Reply:
Why? The reason why we stick to Reno is because that's the only RFC we have. Rather than re-specifying everything here we should reply on what is specified there, so people understand what's congestion control specific (and could be changed by other congestion controllers) and what's a default requirement to implement QUIC.

Reply:
That's not the reason we stick with Reno -- we have RFC 8312 (Cubic) as well. Reno is simple, which is why we went with it. To be clear, this is _not_ a default requirement for QUIC. It's stated that some congestion controller is necessary, and we specify one, but implementations can use others ([Section 7](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#section-7-2).)

Reply:
We're chartered to use a *Standard* congestion controller, and RFC 8312 is Informational.  But as you say, people are perfectly able to plug in other congestion controllers instead; this is merely "a default scheme," in the language of the charter.

Reply:
Thanks for trying to make that clearer. However, the normative language for reduction is most less strict (MUST reduce) and not specifying a concrete algorithm. I think that would be also the better approach for the increase behaviour. However, it's probably less clear what this minimum requirement is. So I would still prefer to just not use normative language here and just say NewReno does increase by  one maximum packet size per congestion window acknowledged.